File: python/helm/session.py
"""Suspendable SE-mode simulation session.

Provides the pause → inspect → hot-load plugin → continue workflow
without requiring the Rust native bindings.  When the native engine
is available it delegates to :class:`helm_engine::SeSession`; otherwise
it runs a pure-Python stub that exercises the same API.

Examples
--------
::

    from helm.session import SeSession
    from helm.plugins import FaultDetect

    s = SeSession("./binary", ["binary"])

    # Phase 1: warm-up (no plugins, full speed)
    s.run(1_000_000)

    # Phase 2: hot-load debug plugin
    s.add_plugin(FaultDetect())

    # Phase 3: run to a specific PC
    reason = s.run_until_pc(0x411120)
    print(f"stopped: {reason}, PC={s.pc:#x}, insns={s.insn_count}")

    # Phase 4: continue with plugin active
    s.run(10_000_000)
    s.finish()
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum, auto
from typing import List, Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from helm.plugins.base import PluginBase


class StopReason(Enum):
    """Why a ``run*`` call returned."""

    INSN_LIMIT = auto()
    BREAKPOINT = auto()
    EXITED = auto()
    ERROR = auto()


@dataclass
class StopResult:
    """Details of why execution stopped."""

    reason: StopReason
    pc: int = 0
    exit_code: int = 0
    message: str = ""

    def __repr__(self) -> str:
        if self.reason == StopReason.EXITED:
            return f"StopResult(EXITED, code={self.exit_code})"
        if self.reason == StopReason.BREAKPOINT:
            return f"StopResult(BREAKPOINT, pc={self.pc:#x})"
        if self.reason == StopReason.ERROR:
            return f"StopResult(ERROR, {self.message!r})"
        return f"StopResult(INSN_LIMIT)"


class SeSession:
    """Suspendable syscall-emulation session.

    Parameters
    ----------
    binary : str
        Path to the AArch64 ELF binary.
    argv : list[str]
        Argument vector (``argv[0]`` is typically the binary name).
    envp : list[str], optional
        Environment variables.
    """

    def __init__(
        self,
        binary: str,
        argv: Optional[List[str]] = None,
        envp: Optional[List[str]] = None,
    ) -> None:
        import os

        self.binary = binary
        self.argv = argv or [os.path.basename(binary)]
        self.envp = envp or [
            "HOME=/tmp", "TERM=dumb", "PATH=/usr/bin:/bin",
            "LANG=C", "USER=helm",
        ]
        self._plugins: List[PluginBase] = []
        self._insn_count: int = 0
        self._virtual_cycles: int = 0
        self._pc: int = 0
        self._exited: bool = False
        self._exit_code: int = 0
        self._native = None

        try:
            self._init_native()
        except ImportError:
            pass

    def _init_native(self) -> None:
        """Try to initialise a Rust-backed session."""
        from helm._helm_core import SeSession as _NativeSeSession
        self._native = _NativeSeSession(self.binary, self.argv, self.envp)

    # -- Plugin management -----------------------------------------------

    def add_plugin(self, plugin: "PluginBase") -> None:
        """Hot-load a plugin.  Takes effect on the next ``run*`` call.

        Can be called before the first run (equivalent to always-on)
        or between runs (the pause/load/continue pattern).
        """
        self._plugins.append(plugin)

    @property
    def plugins(self) -> "List[PluginBase]":
        return list(self._plugins)

    # -- Execution -------------------------------------------------------

    def run(self, max_insns: int = 10_000_000) -> StopResult:
        """Execute up to *max_insns* instructions, then return."""
        if self._native is not None:
            result = self._native.run(max_insns)
            self._pc = self._native.pc
            self._insn_count = self._native.insn_count
            self._virtual_cycles = self._native.virtual_cycles
            self._exited = self._native.has_exited
            if self._exited:
                self._exit_code = self._native.exit_code
                return StopResult(StopReason.EXITED, exit_code=self._exit_code)
            return StopResult(StopReason.INSN_LIMIT, pc=self._pc)
        return self._run_stub(max_insns, pc_break=None)

    def run_until_insns(self, total: int) -> StopResult:
        """Run until *total* instructions have executed since session start."""
        remaining = total - self._insn_count
        if remaining <= 0:
            return StopResult(StopReason.INSN_LIMIT)
        return self._run_stub(remaining, pc_break=None)

    def run_until_pc(
        self,
        target_pc: int,
        max_insns: int = 100_000_000,
    ) -> StopResult:
        """Run until PC equals *target_pc* (with a safety limit)."""
        return self._run_stub(max_insns, pc_break=target_pc)

    # -- Inspection ------------------------------------------------------

    @property
    def pc(self) -> int:
        return self._pc

    @property
    def insn_count(self) -> int:
        return self._insn_count

    @property
    def virtual_cycles(self) -> int:
        return self._virtual_cycles

    @property
    def has_exited(self) -> bool:
        return self._exited

    @property
    def exit_code(self) -> int:
        return self._exit_code

    # -- Teardown --------------------------------------------------------

    def finish(self) -> None:
        """Call ``atexit`` on all loaded plugins."""
        for p in self._plugins:
            p.atexit()

    # -- Stub implementation ---------------------------------------------

    def _run_stub(
        self,
        budget: int,
        pc_break: Optional[int],
    ) -> StopResult:
        """Pure-Python stub — simulates the API without executing code.

        When the native engine is available this path is replaced by
        delegation to the Rust ``SeSession``.
        """
        self._insn_count += budget
        self._virtual_cycles += budget
        return StopResult(StopReason.INSN_LIMIT)


class FsSession:
    """Suspendable full-system simulation session.

    Parameters
    ----------
    kernel : str
        Path to the kernel image.
    machine : str
        Machine type (default: "virt").
    append : str
        Kernel command line.
    sysmap : str, optional
        Path to System.map for symbol resolution.

    Examples
    --------
    ::

        from helm.session import FsSession

        s = FsSession("vmlinuz-rpi", machine="virt",
                       append="earlycon=pl011,0x09000000 console=ttyAMA0")
        s.run(100_000_000)
        print(f"PC={s.pc:#x}, insns={s.insn_count}")
        s.run_until_symbol("start_kernel")
    """

    def __init__(
        self,
        kernel: str,
        machine: str = "virt",
        append: str = "",
        sysmap: Optional[str] = None,
    ) -> None:
        self.kernel = kernel
        self.machine = machine
        self.append = append
        self._native = None

        try:
            from helm._helm_core import FsSession as _NativeFsSession
            self._native = _NativeFsSession(kernel, machine, append, sysmap)
        except ImportError:
            pass

    def run(self, max_insns: int = 100_000_000) -> StopResult:
        """Execute up to *max_insns* instructions, then return."""
        if self._native is not None:
            result = self._native.run(max_insns)
            return StopResult(StopReason.INSN_LIMIT, pc=self.pc)
        return StopResult(StopReason.INSN_LIMIT)

    def run_until_symbol(self, sym: str, max_insns: int = 500_000_000) -> StopResult:
        """Run until a named symbol is reached."""
        if self._native is not None:
            result = self._native.run_until_symbol(sym, max_insns)
            return StopResult(StopReason.BREAKPOINT, pc=self.pc)
        return StopResult(StopReason.INSN_LIMIT)

    def run_until_pc(self, target: int, max_insns: int = 500_000_000) -> StopResult:
        """Run until PC equals target."""
        if self._native is not None:
            result = self._native.run_until_pc(target, max_insns)
            return StopResult(StopReason.BREAKPOINT, pc=self.pc)
        return StopResult(StopReason.INSN_LIMIT)

    @property
    def pc(self) -> int:
        if self._native is not None:
            return self._native.pc
        return 0

    @property
    def insn_count(self) -> int:
        if self._native is not None:
            return self._native.insn_count
        return 0

    def xn(self, n: int) -> int:
        """Read general-purpose register Xn."""
        if self._native is not None:
            return self._native.xn(n)
        return 0

    def regs(self) -> dict:
        """Return all registers as a dict."""
        if self._native is not None:
            return self._native.regs()
        return {}

File: crates/helm-engine/src/se/session.rs
//! Suspendable SE-mode session.
//!
//! [`SeSession`] owns all simulation state and exposes incremental
//! execution methods so callers can pause, inspect, hot-load plugins,
//! and resume.
//!
//! ```text
//! let mut s = SeSession::new("./binary", &["binary"], &[])?;
//! s.run_until_insns(1_000_000);        // warm-up without plugins
//! s.add_plugin("fault-detect", "");     // hot-load
//! s.run_until_pc(0x411120);             // run to entry
//! s.run(10_000_000);                    // continue with plugin active
//! println!("{:?}", s.result());
//! ```

use crate::loader;
use crate::loader::TlsInfo;
use crate::monitor::MonitorTarget;
use crate::se::backend::ExecBackend;
use crate::se::linux::{exec_interp, exec_tcg};
use crate::se::thread::Scheduler;
use crate::symbols::SymbolTable;
use helm_core::HelmError;
use helm_isa::arm::aarch64::Aarch64Cpu;
use helm_memory::address_space::AddressSpace;
use helm_plugin::api::ComponentRegistry;
use helm_plugin::api::PluginArgs;
use helm_plugin::runtime::PluginComponentAdapter;
use helm_plugin::PluginRegistry;
use helm_syscall::Aarch64SyscallHandler;
use helm_timing::TimingModel;

/// Why `run*` returned control to the caller.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum StopReason {
    /// Reached the instruction budget.
    InsnLimit,
    /// Hit the requested PC breakpoint.
    Breakpoint { pc: u64 },
    /// Guest called exit.
    Exited { code: u64 },
    /// An unrecoverable error occurred.
    Error(String),
}

/// Suspendable SE-mode simulation session.
///
/// Owns CPU, memory, syscall handler, scheduler, timing, backend,
/// and the plugin registry — everything needed to pause and resume.
pub struct SeSession {
    cpu: Aarch64Cpu,
    mem: AddressSpace,
    syscall: Aarch64SyscallHandler,
    sched: Scheduler,
    tls_info: Option<TlsInfo>,
    backend: ExecBackend,
    plugin_reg: PluginRegistry,
    comp_reg: ComponentRegistry,
    adapters: Vec<PluginComponentAdapter>,
    timing: Box<dyn TimingModel>,
    insn_count: u64,
    virtual_cycles: u64,
    exited: bool,
    exit_code: u64,
    symbols: SymbolTable,
}

impl SeSession {
    /// Create a new session by loading a binary.
    pub fn new(binary_path: &str, argv: &[&str], envp: &[&str]) -> Result<Self, HelmError> {
        Self::with_timing(
            binary_path,
            argv,
            envp,
            Box::new(helm_timing::model::FeModel),
        )
    }

    /// Create a session with a specific timing model.
    pub fn with_timing(
        binary_path: &str,
        argv: &[&str],
        envp: &[&str],
        timing: Box<dyn TimingModel>,
    ) -> Result<Self, HelmError> {
        let loaded = loader::load_elf(binary_path, argv, envp)?;
        let tls_info = loaded.tls_info.clone();
        let mut mem = loaded.address_space;
        let mut cpu = Aarch64Cpu::new();
        cpu.set_se_mode(true);
        cpu.regs.pc = loaded.entry_point;
        cpu.regs.sp = loaded.initial_sp;
        mem.map(0, 0x1000, (true, false, false));

        let mut syscall = Aarch64SyscallHandler::new();
        syscall.set_brk(loaded.brk_base);
        syscall.binary_path = std::fs::canonicalize(binary_path)
            .unwrap_or_else(|_| std::path::PathBuf::from(binary_path))
            .to_string_lossy()
            .into_owned();
        mem.map(loaded.brk_base, 0x1000, (true, true, false));

        let sched = Scheduler::new(cpu.regs.clone(), 1000);

        let mut comp_reg = ComponentRegistry::new();
        helm_plugin::runtime::register_builtins(&mut comp_reg);

        // Extract symbols from the ELF binary
        let elf_data = std::fs::read(binary_path).unwrap_or_default();
        let symbols = SymbolTable::from_elf(&elf_data);

        Ok(Self {
            cpu,
            mem,
            syscall,
            sched,
            tls_info,
            backend: ExecBackend::interpretive(),
            plugin_reg: PluginRegistry::new(),
            comp_reg,
            adapters: Vec::new(),
            timing,
            insn_count: 0,
            virtual_cycles: 0,
            exited: false,
            exit_code: 0,
            symbols,
        })
    }
    /// Switch execution backend (interpretive → TCG or vice-versa).
    pub fn set_backend(&mut self, backend: ExecBackend) {
        self.backend = backend;
    }

    /// Hot-load a plugin by name.  Can be called between `run*` calls.
    ///
    /// Returns `true` if the plugin was found and installed.
    pub fn add_plugin(&mut self, name: &str, args: &str) -> bool {
        let fqn = resolve_plugin_name(name);
        let plugin_args = if args.is_empty() {
            PluginArgs::new()
        } else {
            PluginArgs::parse(args)
        };
        match self.comp_reg.create(&fqn) {
            Some(comp) => {
                let raw = Box::into_raw(comp);
                let mut adapter = unsafe { *Box::from_raw(raw as *mut PluginComponentAdapter) };
                adapter.install(&mut self.plugin_reg, &plugin_args);
                self.adapters.push(adapter);
                true
            }
            None => false,
        }
    }

    /// Run up to `n` instructions, then return.
    pub fn run(&mut self, max_insns: u64) -> StopReason {
        self.run_inner(max_insns, None)
    }

    /// Run until `total` instructions have executed since session start.
    pub fn run_until_insns(&mut self, total: u64) -> StopReason {
        if self.insn_count >= total {
            return StopReason::InsnLimit;
        }
        self.run_inner(total - self.insn_count, None)
    }

    /// Run until PC equals `target` (or up to `max_insns` as a safety limit).
    pub fn run_until_pc(&mut self, target: u64, max_insns: u64) -> StopReason {
        self.run_inner(max_insns, Some(target))
    }

    /// Current guest PC.
    pub fn pc(&self) -> u64 {
        self.cpu.regs.pc
    }

    /// Total instructions executed so far.
    pub fn insn_count(&self) -> u64 {
        self.insn_count
    }

    /// Total virtual cycles so far.
    pub fn virtual_cycles(&self) -> u64 {
        self.virtual_cycles
    }

    /// Whether the guest has exited.
    pub fn has_exited(&self) -> bool {
        self.exited
    }

    /// Guest exit code (valid only if `has_exited()`).
    pub fn exit_code(&self) -> u64 {
        self.exit_code
    }

    /// Run until a named symbol is reached.
    pub fn run_until_symbol(&mut self, sym: &str, max_insns: u64) -> StopReason {
        match self.symbols.lookup(sym) {
            Some(addr) => self.run_until_pc(addr, max_insns),
            None => StopReason::Error(format!("symbol not found: {sym}")),
        }
    }

    /// Read a general-purpose register.
    pub fn xn(&self, n: u32) -> u64 {
        self.cpu.xn(n as u16)
    }

    /// Read memory at an address.
    pub fn read_memory(&self, addr: u64, size: usize) -> Option<Vec<u8>> {
        let mut buf = vec![0u8; size];
        let mem_ptr = &self.mem as *const AddressSpace as *mut AddressSpace;
        unsafe {
            match (*mem_ptr).read(addr, &mut buf) {
                Ok(()) => Some(buf),
                Err(_) => None,
            }
        }
    }

    /// Call `atexit` on all loaded plugins.
    pub fn finish(&mut self) {
        for adapter in &mut self.adapters {
            adapter.atexit();
        }
    }

    // ── internal ────────────────────────────────────────────────────

    fn run_inner(&mut self, budget: u64, pc_break: Option<u64>) -> StopReason {
        if self.exited {
            return StopReason::Exited {
                code: self.exit_code,
            };
        }

        let limit = self.insn_count + budget;
        let has_insn_cbs = self.plugin_reg.has_insn_callbacks();
        let plugins: Option<&PluginRegistry> = if self.adapters.is_empty() {
            None
        } else {
            Some(&self.plugin_reg)
        };

        while self.insn_count < limit {
            if let Some(target) = pc_break {
                if self.cpu.regs.pc == target && self.insn_count > 0 {
                    return StopReason::Breakpoint { pc: target };
                }
            }

            self.sched.load_regs(&mut self.cpu.regs);
            self.syscall.set_tid(self.sched.current_tid());

            let step_result = if plugins.is_none()
                && pc_break.is_none()
                && matches!(self.backend, ExecBackend::Interpretive)
            {
                let batch = (limit - self.insn_count).min(4096);
                exec_interp_batch(
                    &mut self.cpu,
                    &mut self.mem,
                    &mut self.syscall,
                    &mut self.sched,
                    self.timing.as_mut(),
                    &mut self.insn_count,
                    &mut self.virtual_cycles,
                    self.tls_info.as_ref(),
                    batch,
                )
            } else {
                match &mut self.backend {
                    ExecBackend::Interpretive => exec_interp(
                        &mut self.cpu,
                        &mut self.mem,
                        &mut self.syscall,
                        &mut self.sched,
                        self.timing.as_mut(),
                        &mut None,
                        plugins,
                        &mut None,
                        has_insn_cbs,
                        &mut self.insn_count,
                        &mut self.virtual_cycles,
                        self.tls_info.as_ref(),
                    ),
                    ExecBackend::Tcg { cache, interp } => exec_tcg(
                        &mut self.cpu,
                        &mut self.mem,
                        &mut self.syscall,
                        &mut self.sched,
                        self.timing.as_mut(),
                        plugins,
                        &mut None,
                        cache,
                        interp,
                        &mut self.insn_count,
                        &mut self.virtual_cycles,
                        self.tls_info.as_ref(),
                    ),
                }
            };

            if let Err(e) = step_result {
                return StopReason::Error(format!("{e}"));
            }

            self.sched.save_regs(&self.cpu.regs);

            if self.syscall.should_exit {
                self.exited = true;
                self.exit_code = self.syscall.exit_code;
                return StopReason::Exited {
                    code: self.exit_code,
                };
            }
        }

        StopReason::InsnLimit
    }
}

impl MonitorTarget for SeSession {
    fn run(&mut self, max_insns: u64) -> StopReason {
        self.run(max_insns)
    }
    fn run_until_pc(&mut self, pc: u64, max_insns: u64) -> StopReason {
        self.run_until_pc(pc, max_insns)
    }
    fn pc(&self) -> u64 {
        self.cpu.regs.pc
    }
    fn xn(&self, n: u32) -> u64 {
        self.cpu.xn(n as u16)
    }
    fn sp(&self) -> u64 {
        self.cpu.current_sp()
    }
    fn read_memory(&self, addr: u64, size: usize) -> Option<Vec<u8>> {
        SeSession::read_memory(self, addr, size)
    }
    fn insn_count(&self) -> u64 {
        self.insn_count
    }
    fn virtual_cycles(&self) -> u64 {
        self.virtual_cycles
    }
    fn current_el(&self) -> u8 {
        self.cpu.regs.current_el
    }
    fn daif(&self) -> u32 {
        self.cpu.regs.daif
    }
    fn sysreg(&self, name: &str) -> Option<u64> {
        match name {
            "nzcv" => Some(self.cpu.regs.nzcv as u64),
            "daif" => Some(self.cpu.regs.daif as u64),
            _ => None,
        }
    }
    fn irq_count(&self) -> u64 {
        0
    }
    fn has_exited(&self) -> bool {
        self.exited
    }
    fn symbols(&self) -> &SymbolTable {
        &self.symbols
    }
}

/// Resolve short plugin name → fully-qualified component type.
fn resolve_plugin_name(short: &str) -> String {
    match short {
        "cache" => "plugin.memory.cache".to_string(),
        "fault-detect" => "plugin.debug.fault-detect".to_string(),
        other => format!("plugin.trace.{other}"),
    }
}
use crate::se::linux::exec_interp_batch;
